openedx/features/journals/api.py

```diff
@@ -0,0 +1,167 @@
+"""
+import logging
+import hashlib
+from urlparse import urlsplit, urlunsplit
+import six
+
+from django.conf import settings
+from django.core.cache import cache
+from django.core.exceptions import ObjectDoesNotExist
+from edx_rest_api_client.client import EdxRestApiClient
+from openedx.core.djangoapps.catalog.models import CatalogIntegration
+from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
+from openedx.core.lib.token_utils import JwtBuilder
+from slumber.exceptions import HttpClientError, HttpServerError
+
+
+LOGGER = logging.getLogger("edx.journals")
+JOURNALS_CACHE_TIMEOUT = 3600  # Value is in seconds
+JOURNALS_API_PATH = '/journal/api/v1/'
+
+# TODO move these to SiteConfiguration
+JOURNALS_API_URL = 'http://journals.app:18606/api/v1/'
+JOURNALS_ROOT_URL = 'http://localhost:18606'
+
+
+class DiscoveryApiClient(object):
+    """
+    Class for interacting with the discovery service journals endpoint
+    """
+    def __init__(self, user):
+        """
+        Initialize an authenticated Discovery service API client by using the
+        provided user.
+        """
+        self.user = user
+        jwt = JwtBuilder(user).build_token([])
+        url = configuration_helpers.get_value('COURSE_CATALOG_API_URL', settings.COURSE_CATALOG_API_URL)
+        self.client = EdxRestApiClient(self.create_journals_url(url), jwt=jwt)
+
+    def create_journals_url(self, url):
+        '''rewrite the discovery url to point to journals endpoint'''
+        split_url = urlsplit(url)
+        override_url = urlunsplit((
+            split_url.scheme,
+            split_url.netloc,
+            JOURNALS_API_PATH,
+            split_url.query,
+            split_url.fragment,
+        ))
+        return override_url
+
+    def get_journals(self, orgs):
+        '''get_journals from discovery, filter on orgs is supplied'''
+        try:
+            if orgs:
+                response = self.client.journals.get(orgs=orgs, status='active')
+            else:
+                response = self.client.journals.get(status='active')
+            LOGGER.info('response is type=%s', type(response))
+            return response.get('results')
+        except (HttpClientError, HttpServerError) as err:
+            LOGGER.exception(
+                'Failed to get journals from discovery-service [%s]',
+                err.content
+            )
+            return []
+
+
+class JournalsApiClient(object):
+    """
+    Class for interacting with the Journals Service
+    """
+    def __init__(self, user):
+        """
+        Initialize an authenticated Enterprise service API client by using the
+        provided user.
+        """
+        self.user = user
+        jwt = JwtBuilder(user).build_token([])
+        self.client = EdxRestApiClient(
+            configuration_helpers.get_value('JOURNALS_API_URL', JOURNALS_API_URL),
+            jwt=jwt
+        )
+
+
+def get_cache_key(**kwargs):
```

Comment:
I was going to say we should use whatever pattern the rest of platform is using, then I looked through the code base and realized it was a mess of custom keys. I like this though, it's clean.